File: ios/RNMarketingCloudSdk.m
// RNMarketingCloudSdk.m
//
// Copyright (c) 2019 Salesforce, Inc
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// Redistributions of source code must retain the above copyright notice, this
// list of conditions and the following disclaimer. Redistributions in binary
// form must reproduce the above copyright notice, this list of conditions and
// the following disclaimer in the documentation and/or other materials
// provided with the distribution. Neither the name of the nor the names of
// its contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

@interface RCT_EXTERN_MODULE (RNMarketingCloudSdk, RCTEventEmitter)

RCT_EXTERN_METHOD(isPushEnabled
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(enablePush)

RCT_EXTERN_METHOD(disablePush)

RCT_EXTERN_METHOD(getSystemToken
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setSystemToken:(NSString *_Nonnull)systemToken
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getDeviceID
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setContactKey : (NSString *_Nonnull)contactKey)

RCT_EXTERN_METHOD(getContactKey
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(addTag
                  : (NSString *_Nonnull)tag
                  resolver: (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(removeTag
                  : (NSString *_Nonnull)tag
                  resolver: (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getTags
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setAttribute : (NSString *_Nonnull)name value : (NSString *_Nonnull)value)

RCT_EXTERN_METHOD(clearAttribute : (NSString *_Nonnull)name)

RCT_EXTERN_METHOD(getAttributes
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(enableVerboseLogging)

RCT_EXTERN_METHOD(logSdkState)

RCT_EXTERN_METHOD(getSdkState
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(track : (NSString *_Nonnull)name attributes : (NSDictionary *_Nonnull)attributes)

RCT_EXTERN_METHOD(refresh
                  : (RCTPromiseResolveBlock)resolve rejecter
                  : (RCTPromiseRejectBlock)reject)

@end

File: ios/RNMarketingCloudSdk.swift
import os

@objc(RNMarketingCloudSdk)
class RNMarketingCloudSdk: RCTEventEmitter, RNMarketingCloudSdkLoggerDelegate {
  
  override class func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  override var methodQueue: DispatchQueue! {
    return DispatchQueue.main
  }
  
  var hasListeners: Bool = false
  var verboseLoggingEnabled: Bool = false
  var logger2: RNMarketingCloudSdkLogger?
  
  override func supportedEvents() -> [String]! {
    return ["onLog"]
  }
  
  override func startObserving() {
    self.hasListeners = true
    self.enableLogging()
  }
  
  override func stopObserving() {
    self.hasListeners = false
  }
  
  func enableLogging() {
    if (self.logger2 != nil) {
      return
    }
    
    self.logger2 = RNMarketingCloudSdkLogger()
    self.logger2?.delegate = self;
    
    SFMCSdk.setLogger(logLevel: .debug, logOutputter: self.logger2!)
  }
  
  func log(message: String) {
    os_log("%@", log: .default, type: .default, message)
  }
  
  func onLog(level: LogLevel, subsystem: String, category: LoggerCategory, message: String) {
    if (self.verboseLoggingEnabled) {
      self.log(message: String(format: "%@ - %@", subsystem, message))
    }
    
    if (self.hasListeners) {
      self.sendEvent(withName: "onLog", body: [
        "level": level.rawValue,
        "subsystem": subsystem,
        "category": category.rawValue,
        "message": message
      ])
    }
  }
  
  
  @objc
  func isPushEnabled(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) -> Void {
    resolve(SFMCSdk.mp.pushEnabled())
  }
  
  @objc
  func enablePush() {
    SFMCSdk.mp.setPushEnabled(true)
  }
  
  @objc
  func disablePush() {
    SFMCSdk.mp.setPushEnabled(false)
  }
  
  @objc
  func getSystemToken(_ resolve: RCTPromiseResolveBlock, rejecter
                      reject: RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.deviceToken())
  }
  
  @objc
  func setSystemToken(_ systemToken: String, resolver resolve
                      : RCTPromiseResolveBlock, rejecter reject
                      : RCTPromiseRejectBlock) {
    do {
      let token = try RNMarketingCloudSdkDeviceToken.init(hexString: systemToken)
      SFMCSdk.mp.setDeviceToken(token.data)
      resolve(true)
    } catch {
      return reject("invalid-hex-token", "Failed to convert token to data type", nil)
    }
  }
  
  @objc
  func getDeviceID( _ resolve
                    : RCTPromiseResolveBlock, rejecter reject
                    : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.deviceIdentifier())
  }
  
  @objc
  func setContactKey(_ contactKey: String) {
    SFMCSdk.identity.setProfileId(contactKey)
  }
  
  @objc
  func getContactKey(_ resolve
                     : RCTPromiseResolveBlock, rejecter reject
                     : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.contactKey())
  }
  
  @objc
  func addTag(_ tag : String, resolver resolve
              : RCTPromiseResolveBlock, rejecter reject
              : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.addTag(tag))
  }
  
  @objc
  func removeTag(_ tag : String, resolver resolve
                 : RCTPromiseResolveBlock, rejecter reject
                 : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.removeTag(tag))
  }
  
  @objc
  func getTags(_ resolve
               : RCTPromiseResolveBlock, rejecter reject
               : RCTPromiseRejectBlock) {
    resolve(Array(arrayLiteral: SFMCSdk.mp.tags()))
  }
  
  @objc
  func setAttribute(_ name: String, value: String) {
    SFMCSdk.identity.setProfileAttribute(name, value)
  }
  
  @objc
  func clearAttribute(_ name: String) {
    SFMCSdk.identity.clearProfileAttribute(key: name)
  }
  
  @objc
  func getAttributes(_ resolve
                     : RCTPromiseResolveBlock, rejecter reject
                     : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.mp.attributes())
  }
  
  @objc
  func enableVerboseLogging() {
    self.enableLogging()
    self.verboseLoggingEnabled = true
  }
  
  @objc
  func logSdkState() {
    self.log(message: SFMCSdk.state())
  }
  
  @objc
  func getSdkState(_ resolve
                   : RCTPromiseResolveBlock, rejecter reject
                   : RCTPromiseRejectBlock) {
    resolve(SFMCSdk.state())
  }
  
  @objc
  func track(_ name: String, attributes: Dictionary<String, Any>) {
    SFMCSdk.track(event: CustomEvent(name: name, attributes: attributes)!)
  }
  
  @objc
  func refresh(_ resolve
               : @escaping RCTPromiseResolveBlock, rejecter reject
               : RCTPromiseRejectBlock) {
    SFMCSdk.mp.refresh { result in
      switch (result) {
      case .noData:
        resolve("throttled");
        break;
        
      case .newData:
        resolve("updated");
        break;
        
      case .failed:
        resolve("failed");
        break;
        
      default:
        resolve("unknown");
      }
    }
  }
}